### PR TITLE
OTEP for dynamic configuration

### PR DESCRIPTION
**Pull request by churanchen**
Will change the filename from 0000-template.md when we're ready to make a pull request to Open Telemetry's otep repo.

A lot of the spec lined up fairly well with the OTEP format, so I copied from it when I thought it appropriate.

I think we should discuss the questions posed here: https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048
What are the security implications and how to guard against them?

I think I also need to dig a bit more to understand the performance implications on the agent.

### CODE REVIEW

File: text/0000-template.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+## Internal details
+
+Our protocol will support this call:
+
+```
+// DynamicConfig is a service on the collector that enables updating metric
+// schedules, trace parameters, and other configurations on an agent/SDK without
+// having to restart the instrumented application. The collector may also act as
+// a bridge between third-party configuration services and the agent/sdk, piping
+// updated configs from a third-party source to an instrumented application.
```
Comment (wtong98) [inline text/0000-template.md]: Maybe we could chop out the doc comments, or put them in the proposal itself? This way, we may be able to reduce some redundancy while keeping the code snippets short

File: text/0000-template.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  bytes last_known_fingerprint = 3;
+}
+```
+
+While the response will look like this:
+
+```
+message ConfigResponse {
+
+  // Optional. The fingerprint associated with this ConfigResponse. Each change
```
Comment (wtong98) [inline text/0000-template.md:50]: Maybe we could elaborate on the fields in the request and response in text rather than copy the code snippet? This way, we won't necessarily be restrained by what's written in the protocal documentation, but can elaborate beyond into its potential application

File: text/0000-template.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+There are performance concerns, given that the agent is periodically reading from a service and using this config to update itself. Everything will be implemented optionally, so this will have minimal impact on users who opt not the dynamically update their configs.
+
+The configuration service can be a potential attack vector for an application instrumented with Open Telemetry, depending on what we allow in the protocol. We can highlight in the protocol comments that we should be cautious about what kind of information the agent divulges in its request as well as the sort of behaviour changes that can come about from a config change. 
+
+## Prior art and alternatives
+
+The alternative is to stick with the status quo, where the agent has a [fixed collection period](https://github.com/open-telemetry/opentelemetry-go/blob/34bd99896311a81cf843475779cae2e1c05e6257/sdk/metric/controller/push/push.go#L72-L76) and a fixed sampling rate.
+
+## Open questions
```
Comment (wtong98) [inline text/0000-template.md:154]: Maybe we can link the original comment?

https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048

File: text/0000-template.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+- What happens if a malicious/accidental config change overwhelms the application/monitoring system? Is it the responsibility of the user to be cautious while making config changes? Should we automatically decrease telemetry exporting if we can detect performance problems?
+
+## Future possibilities
```
Comment (wtong98) [inline text/0000-template.md:158]: Maybe we can also add that the collector can act as a shim between third-party configuration services and the instrumented applications?

File: text/0000-template.md (hunk 1 above)
Comment (churanchen) [reply text/0000-template.md]: I took out some of the comments, but left others. I think it might be more understandable in this format.

File: text/0000-template.md (hunk 2 above)
Comment (churanchen) [reply text/0000-template.md:50]: I think it seems okay to leave field descriptions in the code snipper: https://github.com/open-telemetry/oteps/pull/113/files, and it seems more readable.

We necessarily don't need to be restricted to protocol documentation, and we can add/remove stuff.

File: text/0000-template.md (hunk 4 above)
Comment (churanchen) [reply text/0000-template.md:158]: I think this is already in the internal details section. It's more part of our existing spec, rather than as a future possibility

File: text/0000-template.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+## Explanation
+
+All of this functionality will be optional, so the user does not need to do extra work if they do not need a dynamic config service. If they do, the user, when instrumenting their application, would need to configure the open telemetry agent with the endpoint of their remote configuration service, as well as a Resource and a default config in case we fail to read from the service.
+
+The user must then set up the config service. This can be done throught the collector, which can set up either a stand-alone service, or as something that is an interface to the remote configurations of the user's monitoring and tracing backend, "translating" them to comply with the Open Telemetry configuration protocol
```
Comment (jdmontana) [inline text/0000-template.md]: throught -> through

File: text/0000-template.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
+# A Dynamic Configuration Service for the Agent
```
Comment (SergeyKanzhelev) [inline text/0000-template.md]:
```suggestion
# A Dynamic Configuration Service for the SDK
```

File: text/0000-template.md [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+## Motivation
+
+During normal use, a user could use sparse networking, CPU, and memory resources required to send and store telemetry data by specifying sampling of 0.1% of traces, collecting critical metrics such as SLI-related latency distributions and container CPU stats only every five minutes, and not collecting non-critical metrics. Later, while investigating a production issue, the same user could easily increase information available for debugging by reconfiguring some of their processes to sample 2% of traces, collect critical metrics every minute, and begin collecting metrics that were not deemed useful before. Because this change is centralized and does not require redeploying with new configurations, there is lower friction and risk in updating the configurations.
+
+## Explanation
+
+All of this functionality will be optional, so the user does not need to do extra work if they do not need a dynamic config service. If they do, the user, when instrumenting their application, would need to configure the open telemetry agent with the endpoint of their remote configuration service, as well as a Resource and a default config in case we fail to read from the service.
```
Comment (SergeyKanzhelev) [inline text/0000-template.md]:
```suggestion
All of this functionality will be optional, so the user does not need to do extra work if they do not need a dynamic config service. If they do, the user, when instrumenting their application, would need to configure the open telemetry SDK with the endpoint of their remote configuration service, as well as a Resource and a default config in case we fail to read from the service.
```

Comment (SergeyKanzhelev) [reply text/0000-template.md]: similar change everywhere else...

File: text/0000-template.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+## Trade-offs and mitigations
+
+There are performance concerns, given that the agent is periodically reading from a service and using this config to update itself. Everything will be implemented optionally, so this will have minimal impact on users who opt not the dynamically update their configs.
+
+As mentioned [here](https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048), the configuration service can be a potential attack vector for an application instrumented with Open Telemetry, depending on what we allow in the protocol. We can highlight in the protocol comments that we should be cautious about what kind of information the agent divulges in its request as well as the sort of behaviour changes that can come about from a config change. 
+
+## Prior art and alternatives
+
+The alternative is to stick with the status quo, where the agent has a [fixed collection period](https://github.com/open-telemetry/opentelemetry-go/blob/34bd99896311a81cf843475779cae2e1c05e6257/sdk/metric/controller/push/push.go#L72-L76) and a fixed sampling rate.
```
Comment (SergeyKanzhelev) [inline text/0000-template.md]: this section calls for the prior art in other projects. For example, Prometheus is using pull model which effectively a way to dynamically configure the collection period

Comment (SergeyKanzhelev) [reply text/0000-template.md]: another alternative is to make user re-configure collection period using external module. Basically re-registering all aggregators. BTW, is it a worable solution for GoLang SDK today?

File: text/0000-template.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
+# A Dynamic Configuration Service for the Agent
+
+This proposal adds a configuration service to dynamically configure metric collection and tracing.
```
Comment (SergeyKanzhelev) [inline text/0000-template.md]: be explicit here saying that this proposal mentions tracing, but leaving it as TBD for the next iteration

File: text/0000-template.md [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+
+## Explanation
+
+All of this functionality will be optional, so the user does not need to do extra work if they do not need a dynamic config service. If they do, the user, when instrumenting their application, would need to configure the open telemetry SDK with the endpoint of their remote configuration service, as well as a Resource and a default config in case we fail to read from the service.
+
+The user must then set up the config service. This can be done throught the collector, which can set up either a stand-alone service, or as something that is an interface to the remote configurations of the user's monitoring and tracing backend, "translating" them to comply with the Open Telemetry configuration protocol
```
Comment (jdmontana) [inline text/0000-template.md]: I think we may want to rephrase this for clarity to something like: "This can be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the Open Telemetry configuration protocol."

File: text/0000-template.md [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      // metrics, for which this schedule will apply. The string is used to
+      // match against metric names. It should not exceed 100k characters.
+      message Pattern {
+        oneof match {
+          string equals = 1;       // matches the metric name exactly
+          string starts_with = 2;  // prefix-matches the metric name
+        }
+      }
+
+      // Metrics with names that match a rule in the inclusion_patterns are
```
Comment (jdmontana) [inline text/0000-template.md]: Might be a good idea to specifically say "at least one rule in the inclusion_patterns" instead of "a rule" for clarity. Likewise, "Metrics that match at least one exclusion_patterns". Reduces ambiguity. :)

File: text/0000-template.md [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+  // Optional. The client is suggested to wait this long (in seconds) before
+  // pinging the configuration service again.
+  int32 suggested_wait_time_sec = 5;
+}
+```
+
+On the SDK side, everything will be implemented so that all of this functionality will be optional, and the user is not required to add any new configurations. As stated before, to read from a configuration service, the SDK needs to be configured with the service endpoint, a Resource and a default config. The SDK will periodically read a config from the service. If it fails to do so, it will just use the default config, or the most recent successfully read config. If it reads a new config, it will apply it.
+
+Export frequency from the SDK depends on Schedules. Each Schedule has inclusion_patterns, exclusion_patterns and a CollectionPeriod. Any metrics that match any of the inclusion_patterns and do not match any of the exclusion_patterns will be exported every CollectionPeriod (ie. every minute). The SDK implementation will be changed so that it can export metrics that match one certain pattern at a certain interval, while exporting metrics that match another pattern at a different interval.
```
Comment (jdmontana) [inline text/0000-template.md]: I would rephrase this sentence to "The SDK implementation will be changed so that it can export metrics that match a pattern from one Schedule at that Schedule's collection period, while exporting other metrics that match the patterns of another Schedule at that other Schedule's collection period."

File: text/0000-template.md [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+## Trade-offs and mitigations
+
+There are performance concerns, given that the SDK is periodically reading from a service and using this config to update itself. Everything will be implemented optionally, so this will have minimal impact on users who opt not the dynamically update their configs.
+
+As mentioned [here](https://github.com/open-telemetry/opentelemetry-proto/pull/155#issuecomment-640582048), the configuration service can be a potential attack vector for an application instrumented with Open Telemetry, depending on what we allow in the protocol. We can highlight in the protocol comments that we should be cautious about what kind of information the SDK divulges in its request as well as the sort of behaviour changes that can come about from a config change. 
+
+## Prior art and alternatives
+
+One way to configure metric export schedules is to, instead of pushing the metrics, using a pull mechanism to have metrics pulled whenever wanted. This has been implemented for [Prometheus](https://github.com/open-telemetry/opentelemetry-go/pull/751).
```
Comment (jdmontana) [inline text/0000-template.md]: "use a pull mechanism" -> "use a pull mechanism"

**Review by jdmontana — COMMENTED** [review]
This looks really good! Just a few suggestions.

File: text/0000-template.md [hunk 14]
```diff
@@ -0,0 +1,159 @@
+# A Dynamic Configuration Service for the SDK
+
+This proposal is for an experiment to add a configuration service to dynamically configure metric  periods. Per-metric and tracing configuration is also intended to be added, with details left for a later iteration.
```
Comment (vmingchen) [inline text/0000-template.md]: "configure metric collection periods"

File: text/0000-template.md [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+}
+```
+
+A request to the config service will look like this:
+
+```
+message ConfigRequest{
+
+  // Required. The resource for which configuration should be returned.
+  opentelemetry.proto.resource.v1.Resource resource = 2;
```
Comment (vmingchen) [inline text/0000-template.md]: We removed the version number which use "tag=1". Now let's start from 1. Please also change fingerprint tag to 2.

File: text/0000-template.md [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+      }
+      CollectionPeriod period = 3;
+
+      // Optional. Additional opaque metadata associated with the schedule.
+      // Interpreting metadata is implementation specific.
+      bytes metadata = 4;
+    }
+
+    // For this iteration, since we only want one Schedule that applies to all metrics,
+    // we will have a restriction that schedules must have a length of 0.
```
Comment (vmingchen) [inline text/0000-template.md]: length of 1, and we will not check the patterns when apply the collection period.

File: text/0000-template.md [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+    repeated Schedule schedules = 1;
+
+  }
+  MetricConfig metric_config = 3;
+
+  // Dynamic configs specific to trace, like the sampling rate of a resource.
+  message TraceConfig {
+    // TODO: unimplemented
```
Comment (vmingchen) [inline text/0000-template.md:120]: A little more explanation that tracing is not our focus, but we would like to have a placehold to make sure tracing config can be easily added in the future.

File: text/0000-template.md [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+It is related to [this pull request](https://github.com/open-telemetry/opentelemetry-proto/pull/155)
+
+## Motivation
+
+During normal use, users may wish to collect metrics every 10 minutes. Later, while investigating a production issue, the same user could easily increase information available for debugging by reconfiguring some of their processes to collect metrics every 30 seconds. Because this change is centralized and does not require redeploying with new configurations, there is lower friction and risk in updating the configurations.
+
+## Explanation
+
+This OTEP is a request for an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). It is intended to seek approval to develop a proof of concept. This means no development will be done inside either the Open Telemetry SDK or the collector. Since this will be implemented mostly in the contrib repos, all of this functionality will be optional.
```
Comment (vmingchen) [inline text/0000-template.md]: will be optional. The implementation will be under http://go-contrib and http://.../collector-contrib.

**Review by vmingchen — COMMENTED** [review]
Mostly looks good. Please ask William to take a look as well.

File: text/0000-template.md (hunk 14 above)
Comment (wtong98) [inline text/0000-template.md]:
```suggestion
This proposal is for an experiment to add a configuration service to dynamically configure metric  periods. Per-metric and tracing configurations are also intended to be added, with details left for a later iteration.
```

File: text/0000-template.md [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+The user, when instrumenting their application, can configure the SDK with the endpoint of their remote configuration service, the associated Resource and a default config we revert to if we fail to read from the configuration service.
+
+The user must then set up the config service. This MUST be done through the collector, which can be set up to expose an arbitrary configuration service implementation. Depending on implementation, this allows the collector to either act as a stand-alone configuration service, or as a bridge to remote configurations of the user's monitoring and tracing backend by 'translating' the monitoring backend's protocol to comply with the Open Telemetry configuration protocol.
+
+## Internal details
+
+In the future, it is intended to add per-metric configuration. For example, this would allow the user to collect 5XX error counts ever minute, and CPU usage statistics ever 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.
```
Comment (wtong98) [inline text/0000-template.md]:
```suggestion
In the future, it is intended to add per-metric configuration. For example, this would allow the user to collect 5XX error counts ever minute, and CPU usage statistics every 10 minutes. The remote configuration protocol was designed with this in mind, meaning that it includes more details than simply the metric collection period.
```

Comment (wtong98) [reply text/0000-template.md]: What does "5XX" mean?

File: text/0000-template.md (hunk 15 above)
Comment (wtong98) [reply text/0000-template.md]: Oops! I've gone ahead and restarted the numbering on the protobuf

**Comment by wtong98** [comment]
Had just a few small suggestions, but otherwise looks great!

File: text/0000-template.md (hunk 15 above)
Comment (churanchen) [reply text/0000-template.md]: I've added this to ConfigResponse as well.

File: text/0000-template.md (hunk 19 above)
Comment (churanchen) [reply text/0000-template.md]: Like HTTP server errors. I adjusted it a bit to make it more readable.

File: text/0000-template.md (hunk 16 above)
Comment (churanchen) [reply text/0000-template.md]: Good catch!

File: text/0000-template.md (hunk 14 above)
Comment (churanchen) [reply text/0000-template.md]: Do you think this might be ok? In that the word "configuration" is the concept, and the word "configurations" is the noun.

**Review by vmingchen — APPROVED** [review]
